[BEAM-3157] Generate BeamRecord types from Pojos

akedin opened the pull request:
This implements automatic generation of BeamRecordTypes and BeamRecordSqlTypes from pojo types. Work is being done as part of [BEAM-3157](https://issues.apache.org/jira/browse/BEAM-3157).

Main piece is [RecordFactory](https://github.com/apache/beam/compare/master...akedin:generate-record-types?expand=1#diff-55e6442c81f404c1004a445b550f03c9) which exposes a method to generate BeamRecords from pojos. See [RecordFactoryTest](https://github.com/apache/beam/compare/master...akedin:generate-record-types?expand=1#diff-869b654afa6699d55098a8fc3f2e5740) for usage examples. 

The plan is to integrate this into the Beam SQL framework. Integration into SQL will be done in the future PRs.

Records generation is a major step to simplify conversion of pojo model to BeamRecords. Immediate use case is implementation of Nexmark queries in Beam SQL using existing pojo models. 
This can also be used as a starting point for code generation for schema-aware collections.





Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

akedin commented:
R: @kennknowles @reuvenlax @iemejia 
Please take a look

iemejia commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/FieldValueGetter.java:
```diff
@@ -0,0 +1,33 @@
+ * <p>Implementations of this interface are generated at rutime by {@link RecordFactory}
+ * to map pojo fields to BeamRecord fields.
+ */
+public interface FieldValueGetter<T> {
```
I am a bit puzzled by this Interface, is this really needed ? I mean with the Pojo object we have all that is needed,  we have the name/type by reflection and the value just by invoking the method.

akedin replied:
If you use reflection, then you probably don't need this specific interface. But I don't think that getting rid of it, or switching to reflection will reduce complexity a lot without sacrificing something.

This is how I understand it:

Reflection or other implementation details aside, there are 2 problems:
 - map a pojo class to a record type;
 - given a pojo instance, create a record from it, populating the fields in the order they are specified in the record type;

Creating a record type is straightforward, you need to map pojo fields/getters to record field names and coders. You can do it either by convention or using an explicit spec. Record types can be either cached for pojo classes, or re-created for every pojo instance. Overall this feels straightforward. 

Second problem creating a record from a pojo conceptually looks like this:
 - look up or create the record type for the pojo class;
 - get the fields in correct order (same as in record type);
 - for each field invoke the corresponding getter on the pojo;

This means that now we need a mechanism to map record field types to pojo getters (here I assume that even with reflection we will still need to cache `Method`s).

Given the above you seem to need few abstractions no matter whether you use reflection or not:
- [main component](https://github.com/apache/beam/pull/4204/files#diff-55e6442c81f404c1004a445b550f03c9) to create records based on pojos which will:
   - get the record type for a pojo class;
   - while we are creating a record type we know which fields order we're using and which getters correspond to those, so we are also storing a [mapping of ordered list of getters to corresponding record type](https://github.com/apache/beam/pull/4204/files#diff-9b4f7bc45032683f6e0e7d12294057d8);
   - invoke pojo getters in correct order corresponding to record type;
-  component to create record types, which will:
   - map getter names to record field names;
   - map getter return types to coders;
   - there will be 2 flavors of this ([one](https://github.com/apache/beam/pull/4204/files#diff-8910899b54115ac722de323709421db6), [two](https://github.com/apache/beam/pull/4204/files#diff-271ec0f18add17fe3046ec62984ebc89)) because SQL has its custom coders mapping;

So I feel like most of the components in this PR will be needed in some way or another, and the main question is how you map getters to fields and invoke them.

Using reflection:
 - store each getter in a `List<Method>`;
 - invoke a getter with reflection;

Using code generation:
 - generate a class to represent a `Method` call;
 - invoke getter the class directly;

There are pros and cons for both. How I see it:

Reflection:
 - simpler code because you don't have the code generation logic and custom class to represent the getter method;
 - the implementation is java runtime specific on multiple levels;
 - feels like it might require more serialization code if we want to pass the schema information around;
 - easy to shoot yourself in the foot if you start doing more reflection than needed;

Code generation:
 - more code;
 - potentially can be optimized further by JIT;
 - doesn't have to be at run-time;
 - similar interfaces / mechanisms can be implemented in other languages for schema support;

iemejia replied:
Thanks for specifying this, I understand now better the motivation.sorry I forgot to send some points after checking this, I think we agree in the core ideas/goal, my only complain is more API based. Some notes inline:

> Reflection or other implementation details aside, there are 2 problems:
> 
>     map a pojo class to a record type;
>     given a pojo instance, create a record from it, populating the fields in the order they are specified in the record type;

Sure !

You can quickly achieve this by mapping the POJO class into a BeamRecord type. This is just getting the public get* methods and create a LinkedHashMap<String, Coder> where the key is the name of the field and the coder is the given coder for the type.

Caching this mapping seems like an excellent idea because this is a time consuming step. So it should be cached in some sort of Map<Class, BeamRecordType> where the Key is the Pojo Class.

> Second problem creating a record from a pojo conceptually looks like this:
> 
>     look up or create the record type for the pojo class;
>     get the fields in correct order (same as in record type);
>     for each field invoke the corresponding getter on the pojo;
> 
> This means that now we need a mechanism to map record field types to pojo getters (here I assume that even with reflection we will still need to cache Methods).
> 
> Given the above you seem to need few abstractions no matter whether you use reflection or not:

I disagree, you don’t really need extra abstractions for this if you do it by reflection, but more important that needing or not these abstractions is the fact of exposing them in the API. You can do this by simply having a method with a signature like this:

	List<Object> getValues(Object myPojo, List<String> attributes);

and then with that List of values and the BeamRecordType you can build the BeamRecord.

The way you implement this method to make it faster is another story and it is great if we can add some tricks to improve this, but we should try to avoid the optimization from leaking into the API. We are doing this in sdks/java/core, trying not to leak abstractions at this level is important because if we don’t take care here they will be used everywhere.

akedin replied:
I am not sure I understand your point about exposing the abstractions in the API. The only thing I am planning to expose is `BeamRecordFactory.create(pojo)`. This currently creates the `BeamRecord`s using all public getters from the pojo instances. This PR scope is to only implement this.
Probably eventually there will be another parameter added to specify a schema. Whether it's backed by reflection or code generation should not be leaked at this point. Actual schema support or even getting the `BeamRecord` for a subset of the getters of a pojo is out of the scope here.

Imagine that it is plugged into SQL to convert from pojos to records somewhat like this:

```
class PojoToRecord extends DoFn<Pojo, BeamRecord>() {
  @ProcessElement
  public void processElement(Context c) { 
     c.output(BeamRecordFactory.instance().create(c.element()));
  }
```

> `List<Object> getValues(Object myPojo, List<String> attributes);`

Yes, I agree. Something like this is needed. But to actually create a record you also need to make sure that the `attributes` here and the `BeamRecordType` have the same attributes in the same order. 

To achieve that, it would make sense to store that order somewhere, like `Map<Class, List<String>>`, and construct it at the same time you create the `BeamRecordType` because you're analyzing getters anyway. This way the list of attributes is available in the map when you are creating the beam records later. Then you use this list of attributes to get the values by calling `getValues(myPojo, attributes)`. 

I agree it's all pretty straightforward here, but I argue that it's roughly what's happening here anyway, minus the actual code generation bit.

For example, consider following:
- you can combine the values from `Map<Class, BeamRecordType>` and `Map<Class, List<String>>`  and put them in the utility class: `RecordTypeGetters { beamRecordType, attributes }`, and have a single map `Map<Class, RecordTypeGetters>`.
- when calling `getValues(myPojo, attributes)` the most expensive part is actually looking up the methods on the class, so I would cache them:
  - instead of storing the attribute names, just store the getter methods in a `List<Method>`;
  - so now you can change `getValues(Object myPojo, List<String> attributes)` to `getValues(Object myPojo, List<Method> getters)`;
  - now something like this is already in `BeamRecordFactory`: `List<Object> getFieldValues(List<FieldValueGetter> fieldValueGetters, Object pojo)`;
- you still need to wrap everything into some utility/factory class:
  - getting the getters from pojo;
  - mapping getter return types to coders for sql and non-sql versions;
  - caching;
  - getting the values from pojo;
 - now if you replace super low-level `Method` with our own class `FieldValueGetter` you have my implementation here. The only thing different is whether we get `FieldValueGetter`/`Method` by reflection or by codegen;

lukecwik commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/DefaultRecordTypeFactory.java:
```diff
@@ -0,0 +1,91 @@
+ */
+public class DefaultRecordTypeFactory implements RecordTypeFactory {
+
+  private static final ImmutableMap<Class, Coder> JAVA_TYPES_CODERS = ImmutableMap
```
+1 on using the registry to figure this out.

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/RecordFactory.java:
```diff
@@ -0,0 +1,108 @@
+   * <p>Field names for getters are stripped of the 'get' prefix.
+   * For example record field 'name' will be generated for 'getName()' pojo method.
+   */
+  public BeamRecord newRecordCopyOf(Object pojo) {
```
I wanted to make clear it's a copy, not a pass-through wrapper backed by the original pojo

iemejia replied:
create() then ?

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/RecordFactory.java:
```diff
@@ -0,0 +1,108 @@
+   * For example record field 'name' will be generated for 'getName()' pojo method.
+   */
+  public BeamRecord newRecordCopyOf(Object pojo) {
+    RecordTypeGetters getters = getRowType(pojo.getClass());
```
Ah, yes, these are typos, will double check to use `Record` everywhere. I don't have a strong opinion whether to use Row or Record, so will continue to use Record for now until there's a decision to rename to Rows.

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/DirectAccessGetterFactory.java:
```diff
@@ -0,0 +1,119 @@
+ * See {@link ByteBuddyUtils#makeNewGetterInstance(String, DynamicType.Builder)}
+ * and ByteBuddy documentation for details.
+ */
+public class DirectAccessGetterFactory {
```
Main reason is that current Nexmark model has data objects only with public fields. It's relatively easy to convert them to getters but I didn't see any harm in supporting this as well.

iemejia replied:
I see, in this case we will change Nexmark to have public getters, this implementation should stick to the POJO 'semantics' and avoid to move from it (to avoid possible issues).

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/RecordFactory.java:
```diff
@@ -0,0 +1,108 @@
+ *
+ * <p>Currently only 1-1 mapping is supported between pojo classes and record types.
+ */
+public class RecordFactory {
```
<!--thread_id:cc_154609288_t; commit:12389cee6a5061c472533ec063d041e03007687b; resolved:1-->
<!--section:context-quote-->
> **iemejia** wrote:
> BeamRecordFactory ? Probably to remind us in the future to rename this when we move BeamRecord -> Row (or Record).

<!--section:body-->
Sure, renamed this one as it is the main interface. Kept other classes without `Beam` prefix for now. Let me know if you think other ones would be also better with `Beam`

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/DefaultRecordTypeFactory.java:
```diff
@@ -0,0 +1,91 @@
+ */
+public class DefaultRecordTypeFactory implements RecordTypeFactory {
+
+  private static final ImmutableMap<Class, Coder> JAVA_TYPES_CODERS = ImmutableMap
```
<!--thread_id:cc_154606380_t; commit:12389cee6a5061c472533ec063d041e03007687b; resolved:1-->
<!--section:context-quote-->
> **lukecwik** wrote:
> +1 on using the registry to figure this out.

<!--section:body-->
Changed to use `CoderRegistry.createDefault()`. Will update this to get the registry from the pipeline as soon as wire it up (future PRs).

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/DefaultRecordTypeFactory.java:
```diff
@@ -0,0 +1,91 @@
+   * Takes field names and types from the getters, maps them to coders, returns
+   * an instance of the {@link BeamRecordType}.
+   *
+   * <p>Coders not for all field types are implemented.
```
<!--thread_id:cc_154606673_t; commit:12389cee6a5061c472533ec063d041e03007687b; resolved:1-->
<!--section:context-quote-->
> **iemejia** wrote:
> This line is hard to read, maybe remove.

<!--section:body-->
Updated

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/FieldValueGetter.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * An interface to access a field of a class.
+ *
+ * <p>Implementations of this interface are generated at rutime by {@link RecordFactory}
```
<!--thread_id:cc_154598391_t; commit:12389cee6a5061c472533ec063d041e03007687b; resolved:1-->
<!--section:context-quote-->
> **iemejia** wrote:
> s/rutime/runtime

<!--section:body-->
Fixed

akedin commented on sdks/java/core/src/test/java/org/apache/beam/sdk/values/reflect/RecordFactoryTest.java:
```diff
@@ -0,0 +1,132 @@
+import org.junit.Test;
+
+/**
+ * Unit tests for {@link RecordFactoryTest}.
```
<!--thread_id:cc_154598574_t; commit:12389cee6a5061c472533ec063d041e03007687b; resolved:1-->
<!--section:context-quote-->
> **iemejia** wrote:
> s/RecordFactoryTest/RecordFactory

<!--section:body-->
Fixed

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/RecordFactory.java:
```diff
@@ -0,0 +1,108 @@
+ *
+ * <p>Generated record types are cached in the instance of this factory.
+ *
+ * <p>Currently only 1-1 mapping is supported between pojo classes and record types.
```
<!--thread_id:cc_154608554_t; commit:12389cee6a5061c472533ec063d041e03007687b; resolved:1-->
<!--section:context-quote-->
> **iemejia** wrote:
> Maybe add that only 'primitive' types are supported because it does not support more complex types, Lists/Sets/Maps, etc.

<!--section:body-->
This class doesn't explicitly limit what types are supported. It uses `RecordTypeFactory` to create record types. After updating `DefaultRecordTypeFactory` to use `CoderRegistry` it should allow using a lot more types, e.g. byte arrays, sets, serializables. Not sure what would be the use of it at the moment, because it's not used in SQL. `SqlRecordTypeFactory` is the one used for SQL and it does not support complex types at the moment.

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/DirectAccessGetterFactory.java:
```diff
@@ -0,0 +1,119 @@
+ * See {@link ByteBuddyUtils#makeNewGetterInstance(String, DynamicType.Builder)}
+ * and ByteBuddy documentation for details.
+ */
+public class DirectAccessGetterFactory {
```
<!--thread_id:cc_154707090_t; commit:12389cee6a5061c472533ec063d041e03007687b; resolved:1-->
<!--section:context-quote-->
> **iemejia** wrote:
> I see, in this case we will change Nexmark to have public getters, this implementation should stick to the POJO 'semantics' and avoid to move from it (to avoid possible issues).

<!--section:body-->
Removed the direct field accessor implementation.

Although my understanding is that Pojo doesn't imply getter-only. Beans imply getters, but Pojos only imply independence, e.g. no interfaces implementation, custom annotations etc.

akedin commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/RecordFactory.java:
```diff
@@ -0,0 +1,108 @@
+   * <p>Field names for getters are stripped of the 'get' prefix.
+   * For example record field 'name' will be generated for 'getName()' pojo method.
+   */
+  public BeamRecord newRecordCopyOf(Object pojo) {
```
<!--thread_id:cc_154609745_t; commit:12389cee6a5061c472533ec063d041e03007687b; resolved:1-->
<!--section:context-quote-->
> **iemejia** wrote:
> create() then ?

<!--section:body-->
done

kennknowles commented:
If I understand correctly, we may want to make different choices about generating bytecode (and how to do it) versus cached reflection (and how to do this) so the best thing to do would be to put the interface in a place that allows us to do each of these things and compare their performance. We can then change our mind over time.

Is there something in the API that would be a problem for this?

akedin commented:
please retest this

akedin commented:
retest this please

akedin commented:
retest this please

kennknowles commented:
Reviewed 12 of 18 files at r1, 12 of 12 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java, line 53 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EaapDvP4fffX3Ri73:-L3EaapEwKRgdDuP2BIL:b-axovi2) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java#L53)):*
> ```Java
>    * <p>Use this to create instances of {@link BeamRecordType}.
>    */
>   public BeamRecordFactory() {
> ```

Recommended pattern for classes like this: A single private constructor that takes all the arguments. Then for each way of building one, have a `public static` method with a useful name. So this constructor would be `default()` and the one below would be `forRecordTypeFactory(...)`, or something like that.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java, line 81 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3Eb11oH6FUBuXqfPWb:-L3Eb11oH6FUBuXqfPWc:b7m6fwt) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java#L81)):*
> ```Java
>     RecordTypeGetters getters = getRecordType(pojo.getClass());
>     List<Object> fieldValues = getFieldValues(getters.valueGetters(), pojo);
>     return new BeamRecord(getters.recordType(), fieldValues);
> ```

Incidentally, `BeamRecord` should also have clear named factory methods instead of public constructor.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/DefaultRecordTypeFactory.java, line 30 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EbCV00DR3T1uct2YB:-L3EbCV00DR3T1uct2YC:bxbngdc) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/DefaultRecordTypeFactory.java#L30)):*
> ```Java
> 
> /**
>  * A default factory to create a {@link BeamRecordType} based on pojo field getters.
> ```

Can you describe what its behavior is like? Why is it the default? Also - is it `@Internal`?

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/package-info.java, line 22 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EYCva7WmOEvE3Cm3S:-L3EYCvbQTANjyosO0Nx:b-4oyf44) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/package-info.java#L22)):*
> ```Java
>  * Classes to generate BeamRecords from pojos.
>  */
> package org.apache.beam.sdk.values.reflect;
> ```

Is there a special reason to make a new package? It does mean there might be things that are `public` but are not intended for users. So what we do in that case is mark them `@Internal` and also make a big bold comment that says "For internal use only. No backwards compatibility guarantees." Of if things that are public are what you want and you made stuff package private where appropriate, that's great.

And the other thing to do with the package is to set up static analysis with `@DefaultAnnotation(NonNull.class)`

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/FieldValueGetter.java, line 29 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EZLmdr-9mklDOvgfv:-L3EZLmdr-9mklDOvgfw:b-wq2lyt) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/FieldValueGetter.java#L29)):*
> ```Java
>  * to map pojo fields to BeamRecord fields.
>  */
> public interface FieldValueGetter<T> {
> ```

While this is a pretty harmless interface, I think it should still be `@Internal`. I assume that SQL and the new DoFn code will perhaps interface with it so it will have to be public no matter what package, yea?

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/GeneratedGetterFactory.java, line 69 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EZie0jOzoyGC_VSKx:-L3EZie0jOzoyGC_VSKy:b-ldffyq) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/GeneratedGetterFactory.java#L69)):*
> ```Java
>  * }</pre>
>  *
>  * <p>ByteBuddy is used to generate the code. Class naming is left to ByteBuddy's defaults.
> ```

With the DoFn stuff, we had issue where the stacktrace had pseudo-random bits in it so aggregated errors across workers didn't work. I think these bits are expected to be correct-by-construction and never allowed to throw exceptions, yes?

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/package-info.java, line 22 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3E_3p8Q6D5s1nl-tU5:-L3E_3p8Q6D5s1nl-tU6:bk72um2) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/package-info.java#L22)):*
> ```Java
>  * Classes to access fields of pojos to map to record fields.
>  */
> package org.apache.beam.sdk.values.reflect.field;
> ```

Is there a benefit to even more deeply nesting packages? I'd keep it a bit more flat just to make things easy to find and make it most flexible in terms of managing public/private.

---

*[sdks/java/core/src/test/java/org/apache/beam/sdk/values/reflect/BeamRecordFactoryTest.java, line 39 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3E_cEI8Dj6eyUy6Cc5:-L3E_cEI8Dj6eyUy6Cc6:b-wq97rr) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/test/java/org/apache/beam/sdk/values/reflect/BeamRecordFactoryTest.java#L39)):*
> ```Java
>  * Unit tests for {@link BeamRecordFactory}.
>  */
> @RunWith(Parameterized.class)
> ```

Nice use

---

*[sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamRecordSqlType.java, line 90 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EaA3IceENKUiKC4QU:-L3EaA3IceENKUiKC4QV:b-j5rokk) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamRecordSqlType.java#L90)):*
> ```Java
>   public List<Integer> fieldTypes;
> 
>   public static BeamRecordSqlType create(List<String> fieldNames,
> ```

Is this class for internal use only or also do users interact with it? It should either be `@Internal` and with lots of caveats or should have javadoc on public factor methods like this one. Etc throughout (not actually related to this particular pull request)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4204)*
<!-- Sent from Reviewable.io -->

akedin commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java, line 53 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EaapDvP4fffX3Ri73:-L3Eo8f8HYnSACB9u_cC:b9tm155) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java#L53)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

Recommended pattern for classes like this: A single private constructor that takes all the arguments. Then for each way of building one, have a `public static` method with a useful name. So this constructor would be `default()` and the one below would be `forRecordTypeFactory(...)`, or something like that.
</blockquote></details>

Agree. Hidden the constructors, added `BeamRecordFactory.createDefault()`.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java, line 81 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3Eb11oH6FUBuXqfPWb:-L3EoM6PRJnMwb3aLRTD:b89ektd) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java#L81)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

Incidentally, `BeamRecord` should also have clear named factory methods instead of public constructor.
</blockquote></details>

I agree. But there appears to be more than 70 usages of the constructors. I'd prefer to do this in another PR.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/DefaultRecordTypeFactory.java, line 30 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EbCV00DR3T1uct2YB:-L3EocMgChtEmvJG4VtS:b-6u42jz) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/DefaultRecordTypeFactory.java#L30)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

Can you describe what its behavior is like? Why is it the default? Also - is it `@Internal`?
</blockquote></details>

Updated the javadoc with this.

Short version:

This factory creates record types from the list of getters.

Record types are represented by `BeamRecordType` which essentially is a `List<Pair<FieldName, Coder>>`. Getters are represented by a `List<FieldValueGetter>`. So the factory creates former from the latter.

Normally you can do this by mapping java types of the getters to coders using `CoderRegistry` or even manual mapping. This is what this factory does.

Problem with SQL is that it uses `java.sq.Types` to describe types of the record. It keeps this information in `BeamRecordSqlType` in addition to field names and coders. And it is referenced across a bunch of SQL classes. To map the `java.sql.Types` to coders we then need a custom mapping, which is done in a SQL-specific implementation of this interface.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/package-info.java, line 22 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EYCva7WmOEvE3Cm3S:-L3Eu79-x2YuT282Tsf3:bhjkvuv) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/package-info.java#L22)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

Is there a special reason to make a new package? It does mean there might be things that are `public` but are not intended for users. So what we do in that case is mark them `@Internal` and also make a big bold comment that says "For internal use only. No backwards compatibility guarantees." Of if things that are public are what you want and you made stuff package private where appropriate, that's great.

And the other thing to do with the package is to set up static analysis with `@DefaultAnnotation(NonNull.class)`
</blockquote></details>

The parent `values` package seems to contain mostly value types around `PCollection` with almost no other logic. Adding a ton of reflection and code-generation code there didn't seem right, they would become noise in the `values` package. 

Haven't thought about `@Internal` until now, thanks for suggestion. Updated:

* Combined `org.apache.beam.sdk.values.reflect` and `org.apache.beam.sdk.values.reflect.field`;
* Made everything package-private except:
  * `RecordTypeFactory` interface. Used in SQL. Marked it as `@Internal`;
  * `FieldValueGetter` interface. Used by the above factory in SQL. Marked as `@Internal`;
  * `BeamRecordFactory`. Is the only functional class accessed outside of the package. Marked as `@Internal`;

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/FieldValueGetter.java, line 29 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EZLmdr-9mklDOvgfv:-L3FCCUNVt42--OR1C7O:b8eqibd) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/FieldValueGetter.java#L29)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

While this is a pretty harmless interface, I think it should still be `@Internal`. I assume that SQL and the new DoFn code will perhaps interface with it so it will have to be public no matter what package, yea?
</blockquote></details>

I agree. Updated

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/GeneratedGetterFactory.java, line 69 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EZie0jOzoyGC_VSKx:-L3FCG2EkAZclt5X9lkV:b-tp6to4) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/GeneratedGetterFactory.java#L69)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

With the DoFn stuff, we had issue where the stacktrace had pseudo-random bits in it so aggregated errors across workers didn't work. I think these bits are expected to be correct-by-construction and never allowed to throw exceptions, yes?
</blockquote></details>

Hm, haven't really thought about this. Generated code itself doesn't throw but technically nothing stops the underlying getters from doing so. Do you have any ideas on how to handle this better?

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/package-info.java, line 22 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3E_3p8Q6D5s1nl-tU5:-L3FDiVCQm0KsniJcM58:b-13mbq) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/field/package-info.java#L22)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

Is there a benefit to even more deeply nesting packages? I'd keep it a bit more flat just to make things easy to find and make it most flexible in terms of managing public/private.
</blockquote></details>

Combined with parent package.

---

*[sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamRecordSqlType.java, line 90 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EaA3IceENKUiKC4QU:-L3FDsMK5JGcqclPCx7I:b-g3lvlm) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamRecordSqlType.java#L90)):*
<details><summary><i>Previously, kennknowles (Kenn Knowles) wrote…</i></summary><blockquote>

Is this class for internal use only or also do users interact with it? It should either be `@Internal` and with lots of caveats or should have javadoc on public factor methods like this one. Etc throughout (not actually related to this particular pull request)
</blockquote></details>

Added javadoc.

It's public. Until we have the code generation wired up the only way to create a record is to manually create a `BeamRecordSqlType` and then use it to create records.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4204)*
<!-- Sent from Reviewable.io -->

akedin commented:
This has been open for over 2 months now. @kennknowles , @iemejia do you see a path to get this merged? Otherwise I will close it

kennknowles commented:
Sorry about that. I think we should merge it. Technically, it has only been lying stale for 12 days... If you would just write less code, then I could get through the reviews! :-)

kennknowles commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Any more comments from others? (don't let me LGTM dissuade further debate, please!)

---

Reviewed 1 of 22 files at r1, 3 of 20 files at r2, 20 of 31 files at r3.
Review status: 11 of 21 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java, line 81 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3Eb11oH6FUBuXqfPWb:-L4EfxobO6JIXjifGKbR:b-vgb7na) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/BeamRecordFactory.java#L81)):*
<details><summary><i>Previously, akedin (Anton Kedin) wrote…</i></summary><blockquote>

I agree. But there appears to be more than 70 usages of the constructors. I'd prefer to do this in another PR.
</blockquote></details>

Yes, just a drive-by comment on past code.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/DefaultRecordTypeFactory.java, line 30 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EbCV00DR3T1uct2YB:-L4EgLbfE8mO20nd49qK:bfs11wi) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/DefaultRecordTypeFactory.java#L30)):*
<details><summary><i>Previously, akedin (Anton Kedin) wrote…</i></summary><blockquote>

Updated the javadoc with this.

Short version:

This factory creates record types from the list of getters.

Record types are represented by `BeamRecordType` which essentially is a `List<Pair<FieldName, Coder>>`. Getters are represented by a `List<FieldValueGetter>`. So the factory creates former from the latter.

Normally you can do this by mapping java types of the getters to coders using `CoderRegistry` or even manual mapping. This is what this factory does.

Problem with SQL is that it uses `java.sq.Types` to describe types of the record. It keeps this information in `BeamRecordSqlType` in addition to field names and coders. And it is referenced across a bunch of SQL classes. To map the `java.sql.Types` to coders we then need a custom mapping, which is done in a SQL-specific implementation of this interface.


</blockquote></details>

I see. LGTM at this juncture.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/GeneratedGetterFactory.java, line 69 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EZie0jOzoyGC_VSKx:-L4EfBMOsHXpo4MM98NY:bhhfqk2) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/GeneratedGetterFactory.java#L69)):*
<details><summary><i>Previously, akedin (Anton Kedin) wrote…</i></summary><blockquote>

Hm, haven't really thought about this. Generated code itself doesn't throw but technically nothing stops the underlying getters from doing so. Do you have any ideas on how to handle this better?
</blockquote></details>

This was the fix for DoFnInvoker: https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/StableInvokerNamingStrategy.java

I think we can go with what you have for now anyhow and rethink later if it comes up. Not a big issue for POJOs, etc.

---

*[sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/package-info.java, line 22 at r2](https://reviewable.io:443/reviews/apache/beam/4204#-L3EYCva7WmOEvE3Cm3S:-L4EfZvNL44djKaXc_ik:b-z08xwh) ([raw file](https://github.com/apache/beam/blob/40c4ff167947ec3addaf9f281ff919e80f97ceb4/sdks/java/core/src/main/java/org/apache/beam/sdk/values/reflect/package-info.java#L22)):*
<details><summary><i>Previously, akedin (Anton Kedin) wrote…</i></summary><blockquote>

The parent `values` package seems to contain mostly value types around `PCollection` with almost no other logic. Adding a ton of reflection and code-generation code there didn't seem right, they would become noise in the `values` package. 

Haven't thought about `@Internal` until now, thanks for suggestion. Updated:

* Combined `org.apache.beam.sdk.values.reflect` and `org.apache.beam.sdk.values.reflect.field`;
* Made everything package-private except:
  * `RecordTypeFactory` interface. Used in SQL. Marked it as `@Internal`;
  * `FieldValueGetter` interface. Used by the above factory in SQL. Marked as `@Internal`;
  * `BeamRecordFactory`. Is the only functional class accessed outside of the package. Marked as `@Internal`;

</blockquote></details>

OK, works for me!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/apache/beam/4204#-:-L4EgSNV274gpNhvq0KF:b6wzq0k)*
<!-- Sent from Reviewable.io -->

akedin commented:
retest this please

akedin commented:
tis green!

iemejia commented:
My excuses @akedin I have totally forgot about this one, and somehow I wasn't receiving any notifications. It is really nicer now. Thanks, Great work !

iemejia commented:
@akedin Can you please just resolve the existing conflict and I will merge immediately after.

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
run java gradle precommit

akedin commented:
found this with similar gradle failures: https://github.com/apache/beam/pull/4139

akedin commented:
yay it is green

iemejia commented:
:+1: Finally win the war over jenkins !
